index.src.html

```diff
@@ -318,20 +329,33 @@
     </ul>
   </section>
 
-  <section algorithm="'create a credential' internal method">
-    <dfn for="Credential" method export>\[[Create]](options)</dfn> is called with a
-    {{CredentialCreationOptions}}, and returns either a {{Credential}}, if one can be created, 
-    `null` if no credential was created, or an error if creation fails due to exceptional situations
+  <h5 id="algorithm-create-cred" algorithm>'create a credential' internal method"</h5>
+    <dfn for="Credential" method export>\[[Create]](origin, options)</dfn> is called with
+    with an [=environment settings object/origin=] and a {{CredentialCreationOptions}},
+    and returns either a {{Credential}}, if one can be created,
+    or `null` if no credential was created,
+    or a [=constructCredential=] [=struct=],
+    or an error if creation fails due to exceptional situations
     (for example, incorrect options could produce a {{TypeError}}):
 
     <ul class="algorithm">
       1.  Return `null`.
     </ul>
-  </section>
 
-  Unless otherwise specified, the [=interface objects=] for [=interfaces=] which [=interface/inherit=] from
-  {{Credential}} will alias {{Credential}}'s implementation of these three internal methods. Since that
-  implementation isn't particularly useful, derived interfaces MUST override one or the other.
+  <h6 id="algorithm-construct-cred-struct">constructCredential struct</h6>
+
+    <dfn>constructCredential</dfn> is a [=struct=] whose single [=item=] is a [=constructCredentialCallback=]:
+
+    :   <code><dfn>constructCredentialCallback</dfn></code>
+    ::  A Web IDL {{Function}} [=callback function type=] value, whose purpose is to construct,
```

Comment:
Since the `constructCredential` value is provided by another spec, I think it should actually hold a sequence of steps, for which [=promise-calling=] the steps with no parameters yields a `Promise<Credential>`.

Reply:
hm, tho what is the difference between returning a sequence of steps and returning a Web IDL {{Function}} [=callback function type=] value ?  The sequence of steps will be spec'd in another specification in any case, yes?   Plus, cannot one [promise call](https://www.w3.org/2001/tag/doc/promises-guide/#promise-calling) a [callback function type value](https://heycam.github.io/webidl/#idl-callback-function)?

Reply:
It's definitely not a https://heycam.github.io/webidl/#Function, since that's 
```webidl
callback Function = any (any... arguments);
```
while we need to constrain the argument and return types. I guess it'll work to define a suitable callback function type, though, and say it's that.

```webidl
callback Promise<Credential>CredentialConstructor(object global);
```

I leaned against it in the first place because callback functions are generally used to wrap JS function references (https://heycam.github.io/webidl/#idl-callback-function), but https://heycam.github.io/webidl/#idl-callback-functions does say someone else wants to use them for platform functions too.

Reply:
Ok, Domenic suggested not going through a WebIDL type unless we actually expect implementers to use their WebIDL bindings, and we shouldn't expect that here. We should just return an algorithm (which is a better way to say it than the "steps" I suggested earlier).

Reply:
Ok. how does one return an algorithm?  How do we declare it over in [webauthn](https://w3c.github.io/webauthn/)?  Are there examples of such?

Reply:
I would say something like:

5. Return an algorithm that takes a global object |global| and executes the following steps:
   1. Let |attestationObject| be a new `ArrayBuffer` ...
   2. &lt;Remaining steps from constructCredentialCallback>

Reply:
hm, ok, thx.  so IIUC the returned alg would actually be defined in the spec defining the credential type, webauthn in our case, yes?

Reply:
Yep.

Reply:
This makes sense to me FWIW; routing through WebIDL seems like unnecessary overhead.

Reply:
ok, thx @mikewest